src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,381 @@
+
+            this.pendingTipHashHeight = newTip;
+        }
+
+        /// <summary>Saves pending <see cref="ProvenBlockHeader"/> items to the <see cref="ProvenBlockHeaderRepository"/>.</summary>
+        private async Task SaveAsync()
+        {
+            if (this.PendingCache.Count == 0)
+            {
+                this.logger.LogTrace("(-)[PROVEN_BLOCK_HEADER_CACHE_EMPTY]");
+                return;
+            }
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                using (await this.lockobj.LockAsync().ConfigureAwait(false))
+                {
+                    var pendingHeaders = new List<ProvenBlockHeader>();
+
+                    for (int i = this.CurrentTipHashHeight.Height; i <= this.pendingTipHashHeight.Height; i++)
+                    {
+                        ProvenBlockHeader cachedHeader = this.GetHeaderFromPendingCache(i);
+
+                        if (cachedHeader != null)
+                        {
+                            pendingHeaders.Add(cachedHeader);
+                            this.PendingCache.TryRemove(i, out ProvenBlockHeader removedItem);
+                        }
+                    }
+
+                    if (pendingHeaders.Count > 0)
+                        await this.provenBlockHeaderRepository.PutAsync(pendingHeaders, this.pendingTipHashHeight).ConfigureAwait(false);
+
+                    if (this.PendingCache.Count == 0)
+                        this.pendingTipHashHeight = null;
+                }
+            }           
+        }
+
+        /// <summary>Gets <see cref="ProvenBlockHeader"></see> items from pending cache with specific key if present.</summary>
+        /// <param name="key">Item's key.</param>
+        /// <returns><see cref="ProvenBlockHeader"></see> if cache contains the item, <c>null</c> otherwise.</returns>
+        private ProvenBlockHeader GetHeaderFromPendingCache(int key)
+        {
+            if (this.PendingCache.TryGetValue(key, out ProvenBlockHeader header))
+            {
+                this.logger.LogTrace("(-):{0}", header);
+
+                return header;
+            }
+
+            return null;
+        }
+
+        /// <summary>Gets <see cref="ProvenBlockHeader"></see> items from store cache with specific key if present.</summary>
+        /// <param name="key">Item's key.</param>
+        /// <returns><see cref="ProvenBlockHeader"></see> if cache contains the item, <c>null</c> otherwise.</returns>
+        private ProvenBlockHeader GetHeaderFromStoreCache(int key)
+        {
+            if (this.Cache.TryGetValue(key, out ProvenBlockHeader header))
+            {
+                this.logger.LogTrace("(-):{0}", header);
+
+                return header;
+            }
+
+            return null;
+        }
+
+        /// <summary>Manages store cache size.  Remove 30% of items if the memory cache size has been reached.</summary>
+        private void ManangeCacheSize()
```

Comment:
This method doesn't work. 
`i` here is item's index. But you are going to cache and using it as a key. And for cache key is header height. 
So after some height it will always remove nothing at all.

Reply:
Why do we even need to remove from cache is this not handled by the component?

Reply:
LRU Cache structure is a bit limited on what we want to do here.  Also now we are basing the max cache size on memory not items I'm going to switch over to a concurrent dictionary.  Making sure this method picks up all items to purge.